core/trino-main/src/main/java/io/trino/operator/DynamicFilterSourceOperator.java

```diff
@@ -114,15 +115,20 @@ public DynamicFilterSourceOperatorFactory(
         public Operator createOperator(DriverContext driverContext)
         {
             checkState(!closed, "Factory is already closed");
-            return new DynamicFilterSourceOperator(
-                    driverContext.addOperatorContext(operatorId, planNodeId, DynamicFilterSourceOperator.class.getSimpleName()),
-                    dynamicPredicateConsumer,
-                    channels,
-                    planNodeId,
-                    maxDisinctValues,
-                    maxFilterSize,
-                    minMaxCollectionLimit,
-                    blockTypeOperators);
+            OperatorContext operatorContext = driverContext.addOperatorContext(operatorId, planNodeId, DynamicFilterSourceOperator.class.getSimpleName());
+            if (!dynamicPredicateConsumer.getResultFuture().isDone()) {
```

Comment:
Yes, the intention here was to avoid computing dynamic filtering for newly arriving splits when the dynamic filtering threshold was exhausted by a previous split. It is not needed for pipelined execution.
I wasn't aware that drivers would also be eagerly created, I assumed that even if splits are delivered all at once, we would create driver lazily only when needed.
Do we plan to change the logic of fault tolerant execution at some point to deliver splits in batches or create drivers lazily ? If we do we might be better off keeping it as the added logic is simple and the overhead might not show up unless we benchmark at scale on real setup. If we agree to keep it I can add some unit testing for it, though I don't have strong preference either way.

Reply:
Currently we create "sealed" tasks with all the splits pre-assigned to allow failure recovery. In normal scenario there shouldn't be more than 64 splits assigned per task (64MB split * 64 = 4GB of data). 

>  we would create driver lazily only when needed.

Actually yeah, you are right. The `DriverSplitRunner` is created eagerly. The `Driver` is created within `DriverSplitRunner` lazily. 

https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java#L1086

To be honest I'm not necessarily against the optimization per say. It's just it's a little difficult to follow due to the future involved.

> if (!dynamicPredicateConsumer.getResultFuture().isDone()) {

Have you considered having a simple `boolean/AtomicBoolean` flag at the `OperatorFactory` level and propagate this information (about skipping collection) through it?

Reply:
With updated PR I've gotten rid of future and added a `isDomainCollectionComplete` method to LocalDynamicFilterConsumer